File: idt/src/interfaces/schemas/approval.py
"""approval-gate 요청/응답 DTO.

Design Ref: §4. 프론트 타입(`idt_front/src/types/approval.ts`)과 1:1 대응한다.
"""
from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, Field, field_serializer

# 초안 미리보기 길이 — 목록 응답을 가볍게 유지한다. 전문은 상세에서 본다.
DRAFT_PREVIEW_CHARS = 200


def _utc_iso(value: datetime | None) -> str | None:
    """Check G12 — DB 는 UTC naive 를 저장한다. 그대로 직렬화하면 'Z' 가 없어
    프론트 new Date() 가 **로컬 시각**으로 해석해 KST 기준 9시간 어긋난다.
    UTC 임을 명시해 내려준다."""
    if value is None:
        return None
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.isoformat()


class _UtcDatetimeModel(BaseModel):
    """모든 datetime 필드를 UTC 표시(+00:00)와 함께 직렬화한다."""

    @field_serializer("*", when_used="json")
    def _serialize_datetimes(self, value):
        return _utc_iso(value) if isinstance(value, datetime) else value


class ApprovalItemResponse(_UtcDatetimeModel):
    """목록 1건. 초안은 미리보기로 잘라 보낸다."""

    id: str
    agent_id: str
    agent_name: str | None = None
    tool_id: str
    draft_preview: str
    status: Literal[
        "pending", "approved", "scheduled",
        "executed", "rejected", "expired", "failed",
    ]
    execute_after: datetime | None = None
    expires_at: datetime
    seen_at: datetime | None = None
    created_at: datetime


class ApprovalListResponse(BaseModel):
    data: list[ApprovalItemResponse]
    pagination: dict


class ApprovalDetailResponse(ApprovalItemResponse):
    """상세 — 초안 전문 + 집행 인자."""

    draft: str
    tool_args: dict
    worker_id: str
    decided_by: str | None = None
    decided_at: datetime | None = None
    decision_reason: str | None = None
    executed_at: datetime | None = None
    error_message: str | None = None


class ApprovalDecisionResponse(_UtcDatetimeModel):
    """승인·거절 결과. message 는 화면에 그대로 띄울 수 있는 문구."""

    id: str
    status: str
    execute_after: datetime | None = None
    message: str


class RejectApprovalRequest(BaseModel):
    """거절 사유는 필수다 — 재개 시 에이전트에 주입되므로 빈 값은 무의미."""

    reason: str = Field(..., min_length=1, max_length=1000)


class ApprovalTickResponse(BaseModel):
    """내부 tick 결과 — 관측·모니터링용."""

    claimed_count: int
    executed_count: int
    failed_count: int
    expired_count: int


def to_item(approval, agent_name: str | None = None) -> ApprovalItemResponse:
    return ApprovalItemResponse(
        id=approval.id,
        agent_id=approval.agent_id,
        agent_name=agent_name,
        tool_id=approval.tool_id,
        draft_preview=(approval.draft or "")[:DRAFT_PREVIEW_CHARS],
        status=approval.status,
        execute_after=approval.execute_after,
        expires_at=approval.expires_at,
        seen_at=approval.seen_at,
        created_at=approval.created_at,
    )


def to_detail(approval, agent_name: str | None = None) -> ApprovalDetailResponse:
    return ApprovalDetailResponse(
        **to_item(approval, agent_name).model_dump(),
        draft=approval.draft,
        tool_args=approval.tool_args or {},
        worker_id=approval.worker_id,
        decided_by=approval.decided_by,
        decided_at=approval.decided_at,
        decision_reason=approval.decision_reason,
        executed_at=approval.executed_at,
        error_message=approval.error_message,
    )


class ApprovalGateSettingsResponse(BaseModel):
    """에이전트별 게이트 설정 (approval-gate Check G3)."""

    available: bool  # 카탈로그에서 관리자가 활성화했는가
    enabled: bool  # 이 에이전트에 적용 중인가 (강제 포함)
    is_enforced: bool  # 관리자 강제 — 소유자가 끌 수 없음
    config: dict


class ApprovalGateSettingsRequest(BaseModel):
    """저장할 config. 키 검증은 MiddlewareConfigPolicy 가 한다(미지 키 거부)."""

    mode: str = "always"
    execute_after: str | None = None
    expires_hours: int = 168
    on_expire: str = "expire"
    timezone: str | None = None

    def to_config(self) -> dict:
        data = self.model_dump()
        if data.get("timezone") is None:
            data.pop("timezone")  # 미지정이면 도메인 기본(Asia/Seoul)
        return data

File: idt_front/src/types/approval.ts
// approval-gate: 승인 게이트 타입 (백엔드 src/interfaces/schemas/approval.py 1:1 대응)
// Design Ref: §4 (API), §5 (UI)

export const APPROVAL_STATUSES = [
  'pending',
  'approved',
  'scheduled',
  'executed',
  'rejected',
  'expired',
  'failed',
] as const;

export type ApprovalStatus = (typeof APPROVAL_STATUSES)[number];

/** 사람이 아직 볼 이유가 있는 상태 — 목록 기본 필터 (백엔드 DEFAULT_STATUSES 와 일치) */
export const ACTIVE_APPROVAL_STATUSES: ApprovalStatus[] = [
  'pending',
  'scheduled',
];

export const APPROVAL_STATUS_LABELS: Record<ApprovalStatus, string> = {
  pending: '승인 대기',
  approved: '승인됨',
  scheduled: '집행 예정',
  executed: '집행 완료',
  rejected: '거절됨',
  expired: '만료됨',
  failed: '집행 실패',
};

/** 상태 배지 색 — 디자인 토큰(CLAUDE.md UI 섹션) 범위 안에서 */
export const APPROVAL_STATUS_TONES: Record<ApprovalStatus, string> = {
  pending: 'bg-violet-50 text-violet-600 border-violet-200',
  approved: 'bg-zinc-50 text-zinc-600 border-zinc-200',
  scheduled: 'bg-blue-50 text-blue-600 border-blue-200',
  executed: 'bg-emerald-50 text-emerald-600 border-emerald-200',
  rejected: 'bg-zinc-100 text-zinc-500 border-zinc-300',
  expired: 'bg-zinc-100 text-zinc-400 border-zinc-200',
  failed: 'bg-red-50 text-red-500 border-red-200',
};

/** 백엔드 Design §6.1 에러 코드 — 화면 문구 매핑의 단일 출처 */
export const APPROVAL_ERROR_MESSAGES: Record<string, string> = {
  APPROVAL_NOT_FOUND: '이미 삭제되었거나 존재하지 않는 요청입니다.',
  APPROVAL_FORBIDDEN: '이 요청을 승인할 권한이 없습니다.',
  APPROVAL_NOT_PENDING: '이미 처리된 요청입니다. 목록을 새로고침했습니다.',
  APPROVAL_EXPIRED: '만료된 요청입니다.',
  APPROVAL_INVALID_WINDOW:
    '집행 예정 시각이 만료 시각보다 늦습니다. 에이전트 설정을 확인하세요.',
  APPROVAL_AGENT_CHANGED:
    '에이전트 구성이 변경되어 이어서 진행할 수 없습니다. 집행만 진행할 수 있습니다.',
};

export const APPROVAL_PAGE_SIZE = 20;

export interface ApprovalItem {
  id: string;
  agent_id: string;
  agent_name: string | null;
  tool_id: string;
  draft_preview: string;
  status: ApprovalStatus;
  execute_after: string | null;
  expires_at: string;
  seen_at: string | null;
  created_at: string;
}

export interface ApprovalDetail extends ApprovalItem {
  draft: string;
  tool_args: Record<string, unknown>;
  worker_id: string;
  decided_by: string | null;
  decided_at: string | null;
  decision_reason: string | null;
  executed_at: string | null;
  error_message: string | null;
}

export interface ApprovalListResponse {
  data: ApprovalItem[];
  pagination: { total: number; page: number; size: number };
}

export interface ApprovalDecisionResponse {
  id: string;
  status: ApprovalStatus;
  execute_after: string | null;
  message: string;
}

export interface RejectApprovalRequest {
  reason: string;
}

/** 에이전트별 승인 게이트 설정 (백엔드 Design §3.4 config 스키마) */
export interface ApprovalGateConfig {
  mode: 'always' | 'off';
  execute_after: string | null;
  expires_hours: number;
  on_expire: 'expire';
  /** Check G13: cron 해석 기준 타임존. 미지정이면 서버 기본 Asia/Seoul */
  timezone?: string;
}

/** GET/PUT /api/v1/agents/{id}/approval-gate 응답 (Check G3) */
export interface ApprovalGateSettings {
  /** 관리자가 카탈로그에서 승인 게이트를 활성화했는가 */
  available: boolean;
  /** 이 에이전트에 적용 중인가 (강제 포함) */
  enabled: boolean;
  /** 관리자 강제 — 소유자가 끌 수 없다 */
  is_enforced: boolean;
  config: ApprovalGateConfig;
}

/** 전용 API 가 소유해 일반 미들웨어 목록에서 감추는 타입 (백엔드 SEPARATELY_MANAGED 와 일치) */
export const SEPARATELY_MANAGED_MIDDLEWARE_TYPES: readonly string[] = [
  'approval_gate',
];

export const DEFAULT_GATE_TIMEZONE = 'Asia/Seoul';

export const DEFAULT_APPROVAL_GATE_CONFIG: ApprovalGateConfig = {
  mode: 'always',
  execute_after: null,
  expires_hours: 168,
  on_expire: 'expire',
};

export const GATE_EXPIRES_HOURS_MIN = 1;
export const GATE_EXPIRES_HOURS_MAX = 720;

/** 종료 상태 — 목록에서 액션 버튼을 감출 판정 */
export const isApprovalTerminal = (status: ApprovalStatus): boolean =>
  status === 'executed' ||
  status === 'rejected' ||
  status === 'expired' ||
  status === 'failed';

/** 승인·거절 가능 여부. scheduled 는 이미 결정된 건이라 재결정 대상이 아니다. */
export const isApprovalActionable = (status: ApprovalStatus): boolean =>
  status === 'pending';
